### PR TITLE
GH-40361:  [C++] Make flatbuffers serialization more deterministic

### PR DESCRIPTION
**Pull request by noamross**
### Rationale for this change

This is the start of a PR to address #40361, and in turn #40202, to make metadata in parquet files written by arrow to be identical irrespective of the platform configuration.  This is limited, as platform-specific differences in R or Python versions or compression libraries could still result in differences.

### What changes are included in this PR?

So far I have only made a partial change to part of the metadata serialization.  I need to look at whether other calls to flatbuffers require similar treatment.

### Are these changes tested?

Not yet, this is a draft PR

### Are there any user-facing changes?

No 

* GitHub Issue: #40361

### CODE REVIEW

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 1]
```diff
@@ -477,7 +477,9 @@ static Status GetDictionaryEncoding(FBB& fbb, const std::shared_ptr<Field>& fiel
 
 static KeyValueOffset AppendKeyValue(FBB& fbb, const std::string& key,
                                      const std::string& value) {
-  return flatbuf::CreateKeyValue(fbb, fbb.CreateString(key), fbb.CreateString(value));
+  auto fbb_key_ = fbb.CreateString(key);
+  auto fbb_value_ = fbb.CreateString(value);
```
Comment (kou) [inline cpp/src/arrow/ipc/metadata_internal.cc:481]: Please don't use `_` suffix. We use it only for member variable (not local variable).

**Comment by amoeba** [comment]
Thanks @noamross!

@kou can you please approve all workflows to run?

**Comment by kou** [comment]
Approved!

**Comment by pitrou** [comment]
As discussed on the GH issue, it would be useful to add a test using output from your machine as a reference.
(also, perhaps open a separate PR with just the test to validate that it would fail on CI)

**Comment by amoeba** [comment]
Hi @noamross, would you be interested in adding the test described above to this PR? If not, I could take a shot at it.

**Comment by noamross** [comment]
@amoeba Sorry to open this and abandon it, it just became a very busy month and my C++ is pretty limited.  I'd be happy if you would!

**Comment by amoeba** [comment]
No problem at all @noamross, it's awesome you filed a PR for the issue and it's very much appreciated. I'll work on adding and testing the test.

**Comment by noamross** [comment]
Thanks! I expect the files attached to https://github.com/apache/arrow/issues/40202#issuecomment-1977273164 could serve for a test of whether a generated file is identical to a reference from a different system, but I note that they would change with the R version being used.

**Comment by amoeba** [comment]
I put up a draft PR with a test to exercise this and hopefully watch it fail on CI at https://github.com/apache/arrow/pull/41018. It fails on my amd64 linux machine so I think it's close to what we need here. I'll watch CI on that PR and report back.

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
Things look good locally (testing on macOS-aarch64-clang and Linux-amd64-gcc) but I see an issue with my includes in a few of the above jobs. I'll fix those soon.

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
I've pushed up a fix to address the includes issue and the crossbow jobs looked clear (failures are unrelated and the test we expect to pass does indeed pass).

I've marked this as ready for review. @pitrou do you have time to review this?

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 1 above)
Comment (kou) [reply cpp/src/arrow/ipc/metadata_internal.cc:481]: @amoeba Could you check this comment?

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// GH-40361: Test that Flatbuffer serialization matches a known output
+// byte-for-byte.
+//
+// Our Flatbuffers code should not depend on argument evaluation order as it's
+// undefined (https://en.cppreference.com/w/cpp/language/eval_order) and may
+// lead to unnecessary platform- or toolchain-specific differences in
+// serialization.
+TEST(TestMessageInternal, TestByteIdentical) {
+  FBB fbb;
+  flatbuffers::Offset<flatbuf::Schema> fb_schema;
```
Comment (kou) [inline cpp/src/arrow/ipc/message_internal_test.cc:48]: It seems that the `flatbuf` shortcut is used only here. How about using `org::apache::arrow::flatbuf` directly?

```suggestion
namespace ipc {
namespace internal {

using FBB = flatbuffers::FlatBufferBuilder;

// GH-40361: Test that Flatbuffer serialization matches a known output
// byte-for-byte.
//
// Our Flatbuffers code should not depend on argument evaluation order as it's
// undefined (https://en.cppreference.com/w/cpp/language/eval_order) and may
// lead to unnecessary platform- or toolchain-specific differences in
// serialization.
TEST(TestMessageInternal, TestByteIdentical) {
  FBB fbb;
  flatbuffers::Offset<org::apache::arrow::flatbuf:Schema> fb_schema;
```

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+#include "arrow/memory_pool.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/util/key_value_metadata.h"
+
+namespace arrow {
+
+namespace flatbuf = org::apache::arrow::flatbuf;
+
+namespace ipc {
+namespace internal {
```
Comment (kou) [inline cpp/src/arrow/ipc/message_internal_test.cc:35]: Could you use `namespace arrow::ipc::internal {` style as much as possible?

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 1 above)
Comment (amoeba) [reply cpp/src/arrow/ipc/metadata_internal.cc:481]: Whoops. Will fix.

Comment (amoeba) [reply cpp/src/arrow/ipc/metadata_internal.cc]: Changed in 4cc60d766727fecd807c45ee9df88824a5b80029.

File: cpp/src/arrow/ipc/message_internal_test.cc (hunk 3 above)
Comment (amoeba) [reply cpp/src/arrow/ipc/message_internal_test.cc]: Done in 0ebc472e966795bf9b69963616b3ba7bc91a23c8.

**Comment by amoeba** [comment]
Thanks for taking a look @kou, I accepted all your changes. I feel pretty good about the state of this PR at this point and am not sure we need another review. Let me know what you think.

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      0x08, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x0B, 0x00,
+      0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F, 0x31, 0x5F, 0x76, 0x61, 0x6C, 0x75, 0x65, 0x00,
+      0x05, 0x00, 0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F, 0x31, 0x00, 0x00, 0x00, 0x01, 0x00,
+      0x00, 0x00, 0x14, 0x00, 0x00, 0x00, 0x10, 0x00, 0x14, 0x00, 0x08, 0x00, 0x06, 0x00,
+      0x07, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x01, 0x02, 0x10, 0x00, 0x00, 0x00, 0x1C, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x66, 0x30, 0x00, 0x00, 0x08, 0x00,
+      0x0C, 0x00, 0x08, 0x00, 0x07, 0x00, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+      0x40, 0x00, 0x00, 0x00};
+  Buffer expectedBuffer(expected, len);
```
Comment (kou) [inline cpp/src/arrow/ipc/message_internal_test.cc:78]: Can we compute `len` automatically?

```suggestion
  const uint8_t expected[] = {
      0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0A, 0x00, 0x0E, 0x00, 0x06, 0x00, 0x05, 0x00,
      0x08, 0x00, 0x0A, 0x00, 0x00, 0x00, 0x00, 0x01, 0x04, 0x00, 0x10, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x0A, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x04, 0x00, 0x08, 0x00, 0x0A, 0x00,
      0x00, 0x00, 0x6C, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00,
      0x38, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0xD8, 0xFF, 0xFF, 0xFF, 0x18, 0x00,
      0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x0B, 0x00, 0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F,
      0x32, 0x5F, 0x76, 0x61, 0x6C, 0x75, 0x65, 0x00, 0x05, 0x00, 0x00, 0x00, 0x6B, 0x65,
      0x79, 0x5F, 0x32, 0x00, 0x00, 0x00, 0x08, 0x00, 0x0C, 0x00, 0x04, 0x00, 0x08, 0x00,
      0x08, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x0B, 0x00,
      0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F, 0x31, 0x5F, 0x76, 0x61, 0x6C, 0x75, 0x65, 0x00,
      0x05, 0x00, 0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F, 0x31, 0x00, 0x00, 0x00, 0x01, 0x00,
      0x00, 0x00, 0x14, 0x00, 0x00, 0x00, 0x10, 0x00, 0x14, 0x00, 0x08, 0x00, 0x06, 0x00,
      0x07, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x01, 0x02, 0x10, 0x00, 0x00, 0x00, 0x1C, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x66, 0x30, 0x00, 0x00, 0x08, 0x00,
      0x0C, 0x00, 0x08, 0x00, 0x07, 0x00, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
      0x40, 0x00, 0x00, 0x00};
  Buffer expectedBuffer(expected, sizeof(expected));
```

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/ipc/message_internal_test.cc (hunk 4 above)
Comment (amoeba) [reply cpp/src/arrow/ipc/message_internal_test.cc]: Oh, sure. That's probably better. Thanks.

**Comment by amoeba** [comment]
@github-actions crossbow submit -g cpp

**Comment by amoeba** [comment]
I'll let CI and the crossbow jobs run then merge if all looks good.

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  // Create a simple Schema with just one metadata KVP
+  auto f0 = field("f0", int64());
+  auto f1 = field("f1", int64());
+  std::vector<std::shared_ptr<Field>> fields = {f0, f1};
+  std::shared_ptr<KeyValueMetadata> metadata =
+      KeyValueMetadata::Make({"key_1", "key_2"}, {"key_1_value", "key_2_value"});
+  auto schema = ::arrow::schema({f0}, metadata);
+
+  // Serialize the Schema to a Buffer
+  std::shared_ptr<Buffer> outBuffer;
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message_internal_test.cc:54]: Nit: local variables should be `snake_case`, so `out_buffer` and `expected_buffer`.

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      0x05, 0x00, 0x00, 0x00, 0x6B, 0x65, 0x79, 0x5F, 0x31, 0x00, 0x00, 0x00, 0x01, 0x00,
+      0x00, 0x00, 0x14, 0x00, 0x00, 0x00, 0x10, 0x00, 0x14, 0x00, 0x08, 0x00, 0x06, 0x00,
+      0x07, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x01, 0x02, 0x10, 0x00, 0x00, 0x00, 0x1C, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
+      0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x66, 0x30, 0x00, 0x00, 0x08, 0x00,
+      0x0C, 0x00, 0x08, 0x00, 0x07, 0x00, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+      0x40, 0x00, 0x00, 0x00};
+  Buffer expectedBuffer(expected, sizeof(expected));
+
+  ASSERT_TRUE(expectedBuffer.Equals(*outBuffer));
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message_internal_test.cc:78]: Can use `AssertBufferEqual`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks @noamross , @amoeba ! Just two minor comments, otherwise LGTM.

File: cpp/src/arrow/ipc/CMakeLists.txt [hunk 7]
```diff
@@ -41,6 +41,7 @@ add_arrow_test(feather_test)
 add_arrow_ipc_test(json_simple_test)
 add_arrow_ipc_test(read_write_test EXTRA_LINK_LIBS arrow::flatbuffers)
 add_arrow_ipc_test(tensor_test)
+add_arrow_ipc_test(message_internal_test EXTRA_LINK_LIBS arrow::flatbuffers)
```
[the remark this answers is not in the document]
Comment (amoeba) [reply cpp/src/arrow/ipc/CMakeLists.txt:44]: Good catch. Done in e0969f8b9aa2a8b616cd3759320194c416354970.

File: cpp/src/arrow/ipc/message_internal_test.cc (hunk 5 above)
Comment (amoeba) [reply cpp/src/arrow/ipc/message_internal_test.cc:54]: Thanks, fixed in 5b2afb0a1d33b54ec9e3e1ca61c0651f63d68cf2.

File: cpp/src/arrow/ipc/message_internal_test.cc (hunk 6 above)
Comment (amoeba) [reply cpp/src/arrow/ipc/message_internal_test.cc:78]: Thanks. Fixed in 7bc02d77a7eb6795b3cd4f13831cd38bf74b0ae9.

**Comment by amoeba** [comment]
Thanks @pitrou for the review. I rebased from apache/main, accepted your three changes as-is, and added one extra change I caught while dealing with conflicts on rebase, 4bd0339e596d410bd5ad3a73f06d9294c7a87e7f. In https://github.com/apache/arrow/commit/4c8eff2ec8bae76364fcb28ffac574c1df262705 @kou  tweaked the CMakeList and I think this PR should make the same change as was made there.

PS: After pushing I found the ipc/CMakeLists.txt needed formatting, so that was added here too.

**Comment by amoeba** [comment]
Parquet tests were failing all over on CI and due to needing to bump submodules so I pushed 4ebe53bd21fd1f28ce6babebf6802488283c692c so we could see CI.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/ipc/message_internal_test.cc [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+// Our Flatbuffers code should not depend on argument evaluation order as it's
+// undefined (https://en.cppreference.com/w/cpp/language/eval_order) and may
+// lead to unnecessary platform- or toolchain-specific differences in
+// serialization.
+TEST(TestMessageInternal, TestByteIdentical) {
+  FBB fbb;
+  flatbuffers::Offset<org::apache::arrow::flatbuf::Schema> fb_schema;
+  DictionaryFieldMapper mapper;
+
+  // Create a simple Schema with just one metadata KVP
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message_internal_test.cc:45]:
```suggestion
  // Create a simple Schema with just two metadata KVPs
```

**Review by pitrou — CHANGES_REQUESTED** [review]
It seems there are still spurious `parquet-testing` updates in this PR?

**Review by mapleFU — APPROVED** [review]
By the way, can you switch the edit of parquet-testing back?

**Comment by amoeba** [comment]
Thanks all. The submodule change to help CI pass has been reverted in this PR. I'm merging this now.

File: cpp/src/arrow/ipc/CMakeLists.txt [hunk 9]
```diff
@@ -6,7 +6,7 @@
 # "License"); you may not use this file except in compliance
 # with the License.  You may obtain a copy of the License at
 #
-#   http://www.apache.org/licenses/LICENSE-2.0
+# http://www.apache.org/licenses/LICENSE-2.0
```
Comment (kou) [inline cpp/src/arrow/ipc/CMakeLists.txt:9]: Could you revert this needless white space change?

Comment (amoeba) [reply cpp/src/arrow/ipc/CMakeLists.txt:9]: Sorry about that but thanks for noticing. PR at https://github.com/apache/arrow/pull/41701.

**Pull request merged**